harness/determined/searcher/search_method.py

```diff
@@ -0,0 +1,245 @@
+            "experimentCompleted": self.experiment_completed,
+        }
+
+    def from_dict(self, d: Dict[str, Any]) -> None:
+        self.failures = {uuid.UUID(f) for f in d.get("failures", [])}
+        self.trial_progress = {uuid.UUID(k): v for k, v in d.get("trialProgress", {}).items()}
+        self.trials_closed = {uuid.UUID(t) for t in d.get("trialsClosed", [])}
+        self.trials_created = {uuid.UUID(t) for t in d.get("trialsCreated", [])}
+        self.last_event_id = d.get("lastEventId", 0)
+        self.experiment_id = d.get("experimentId")
+        self.experiment_completed = d.get("experimentCompleted", False)
+
+
+class ExitedReason(Enum):
+    ERRORED = "ERRORED"
+    USER_CANCELED = "USER_CANCELED"
+    INVALID_HP = "INVALID_HP"
+
+    @classmethod
+    def _from_bindings(
+        cls, bindings_exited_reason: bindings.v1TrialExitedEarlyExitedReason
+    ) -> "ExitedReason":
+        if (
+            bindings_exited_reason
+            == bindings.v1TrialExitedEarlyExitedReason.EXITED_REASON_INVALID_HP
+        ):
+            return cls.INVALID_HP
+        if (
+            bindings_exited_reason
+            == bindings.v1TrialExitedEarlyExitedReason.EXITED_REASON_USER_REQUESTED_STOP
+        ):
+            return cls.USER_CANCELED
+        raise RuntimeError(f"Invalid exited reason: {bindings_exited_reason}")
+
+
+class Operation:
+    @abstractmethod
+    def _to_searcher_operation(self) -> bindings.v1SearcherOperation:
+        pass
+
+
+class ValidateAfter(Operation):
+    def __init__(self, request_id: uuid.UUID, length: int) -> None:
+        super().__init__()
+        self.request_id = request_id
+        self.length = length
+
+    def _to_searcher_operation(self) -> bindings.v1SearcherOperation:
+        return bindings.v1SearcherOperation(
+            validateAfter=bindings.v1ValidateAfterOperation(
+                requestId=str(self.request_id),
+                length=str(self.length),
+            )
+        )
+
+
+class Close(Operation):
+    def __init__(self, request_id: uuid.UUID):
+        super().__init__()
+        self.request_id = request_id
+
+    def _to_searcher_operation(self) -> bindings.v1SearcherOperation:
+        return bindings.v1SearcherOperation(
+            closeTrial=bindings.v1CloseTrialOperation(requestId=str(self.request_id))
+        )
+
+
+class Progress(Operation):
+    def __init__(self, progress: float):
+        super().__init__()
+        self.progress = progress
+
+    def _to_searcher_operation(self) -> bindings.v1SearcherOperation:
+        return bindings.v1SearcherOperation(
+            searcherProgress=bindings.v1SearcherProgressOperation(progress=self.progress)
+        )
+
+
+class Shutdown(Operation):
+    def __init__(self) -> None:
+        super().__init__()
+
+    def _to_searcher_operation(self) -> bindings.v1SearcherOperation:
+        return bindings.v1SearcherOperation(shutdown=bindings.v1ShutdownOperation())
```

Comment:
question: why have a shutdown?  Why not just exit 0 and let that be your shutdown?

Reply:
Then the experiment will never end, it will just stay there, supposedly "ACTIVE", without doing anything. Maybe I am missing something but master already had the concept of a shutdown operation, and we decided to use it the way, I think, it was intended.